# HG changeset patch
# Parent a41b6e47ce93904706eda1fd4deb1d08345fe351
optimize wxDataViewCtrl::Expand() by not calling ExpandAncestors() which will
call Expand() on all parent items and this in turn will call ExpandAncestors()
all over again

diff -r a41b6e47ce93 src/generic/datavgen.cpp
--- a/src/generic/datavgen.cpp	Thu Jul 19 15:54:35 2012 +0200
+++ b/src/generic/datavgen.cpp	Thu Jul 19 15:59:58 2012 +0200
@@ -5090,14 +5090,35 @@
 
 void wxDataViewCtrl::Expand( const wxDataViewItem & item )
 {
-    ExpandAncestors( item );
-
-    int row = m_clientArea->GetRowByItem( item );
-    if (row != -1)
+    const wxDataViewModel* model = GetModel();
+
+    // we cannot call ExpandAncestors() here because this function calls Expand() on all the
+    // parent items and this in turn will call ExpandAncestors() again for every item
+    if (!model || !item.IsOk())
+        return;
+
+    wxVector<wxDataViewItem> items;
+
+    // put the item itself in the array
+    items.push_back(item);
+
+    // at first we get all the parents of the item
+    wxDataViewItem parent = model->GetParent(item);
+    while (parent.IsOk())
     {
+        items.push_back(parent);
+        parent = model->GetParent(parent);
+    }
+
+    // then we expand the parents, starting at the root
+    while (!items.empty())
+    {
+        const int row = m_clientArea->GetRowByItem(items.back());
         m_clientArea->Expand(row);
-        InvalidateColBestWidths();
+        items.pop_back();
     }
+
+    InvalidateColBestWidths();
 }
 
 void wxDataViewCtrl::Collapse( const wxDataViewItem & item )
